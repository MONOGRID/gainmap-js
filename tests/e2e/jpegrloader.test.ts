import 'jest-extended'

import { describe, expect, it } from '@jest/globals'
import { toMatchImageSnapshot } from 'jest-image-snapshot'

import { getPage } from './common'

expect.extend({ toMatchImageSnapshot })

const matrix = [
  ['01.jpg'],
  ['02.jpg'],
  ['03.jpg'],
  ['04.jpg'],
  ['05.jpg'],
  ['06.jpg'],
  ['07.jpg'],
  ['08.jpg'],
  ['09.jpg'],
  ['10.jpg'],
  ['pisa-4k.jpg'],
  ['spruit_sunrise_4k.jpg']
  // ['abandoned_bakery_16k.jpg'] // too big to test? snapshot testing fails
]

describe('JPEGRLoader', () => {
  it.each(matrix)('reconstructs an HDR image from %p', async (file) => {
    const { page, pageError, pageLog } = await getPage('base')
    await page.addScriptTag({
      type: 'module',
      url: 'scripts/jpegrloader.js'
    })
    const result = await page.evaluate(`JPEGRLoader('${file}')`) as { width: number, height: number, data: Uint16Array, max: number }

    expect(pageError).not.toBeCalled()
    // expect no calls to page log except the one indicated
    expect(pageLog).not.toBeCalledWith(
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      expect.not.stringMatching(/GPU stall due to ReadPixels/)
    )

    // we receive Arrays because puppeteer can't transfer Uint8Array data
    result.data = Uint16Array.from(result.data)

    expect(result.data).toBeInstanceOf(Uint16Array)
    expect(result.data.length).toBeGreaterThan(4)
    expect(result.max).toMatchSnapshot('max HDR value')
  }, 900000 /* 15 minutes */)
})
