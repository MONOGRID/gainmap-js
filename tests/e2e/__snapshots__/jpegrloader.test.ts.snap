// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JPEGRLoader reconstructs an HDR image from "01.jpg": max HDR value 1`] = `
[
  6.390625,
  6.296875,
  6.265625,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "02.jpg": max HDR value 1`] = `
[
  13.8515625,
  9.9921875,
  7.7734375,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "03.jpg": max HDR value 1`] = `
[
  10.0625,
  8.6875,
  8.75,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "04.jpg": max HDR value 1`] = `
[
  12.0234375,
  6.53515625,
  2.353515625,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "05.jpg": max HDR value 1`] = `
[
  15.9921875,
  16,
  16,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "06.jpg": max HDR value 1`] = `
[
  15.0234375,
  11.65625,
  8.921875,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "07.jpg": max HDR value 1`] = `
[
  16,
  13.6875,
  9.8203125,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "08.jpg": max HDR value 1`] = `
[
  6.85546875,
  6.1171875,
  5.39453125,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "09.jpg": max HDR value 1`] = `
[
  9.796875,
  9.5,
  9.3125,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "10.jpg": max HDR value 1`] = `
[
  7.84375,
  5.421875,
  2.216796875,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "pisa-4k.jpg": max HDR value 1`] = `
[
  8.8984375,
  8.3125,
  6.26953125,
]
`;

exports[`JPEGRLoader reconstructs an HDR image from "spruit_sunrise_4k.jpg": max HDR value 1`] = `
[
  65504,
  65504,
  39456,
]
`;
